poptorch_experimental_addons/sharded.py

```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+from typing import Any
+
+import einops
+import torch
+
+from .collectives import all_gather_cross_replica, all_reduce_cross_replica_sum
+
+
+def rowcolrow_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, replication_factor: int, num_chunks: int = 1
+) -> Any:
+    """
+    Matrix multiplation for row-sharded x column-sharded -> row-sharded tensors
+
+    Gathers the right multiplicand across IPU program replicas
+    """
+    X_local_outer_dim, _ = X.shape
+    _, Y_local_outer_dim = Y.shape
+    result = torch.zeros((X_local_outer_dim, Y_local_outer_dim * replication_factor))
+    X = einops.rearrange(X, "m (k c) -> m k c", k=num_chunks)
+    Y = einops.rearrange(Y, "(k c) n -> c k n", k=num_chunks)
+    for i in range(num_chunks):
+        index = torch.tensor([i])
+        Yg = all_gather_cross_replica(
+            torch.index_select(Y, dim=1, index=index).squeeze(), replication_factor
+        )
+        Yg = einops.rearrange(Yg, "r c n -> c (n r)")
+        Xp = torch.index_select(X, dim=1, index=index).squeeze()
+        result += Xp @ Yg
+    return result
+
+
+def repcolcol_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, replication_factor: int
+) -> Any:
+    """Matrix multiplation for replicated x column-sharded -> column-sharded tensors"""
+    X = all_reduce_cross_replica_sum(X, replication_factor, insert_in_grad_graph=True)
+    return X @ Y
+
+
+def colrowrep_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, replication_factor: int
+) -> Any:
+    """Matrix multiplation for row-sharded x column-sharded -> replicated tensors"""
+    out = X @ Y
+    return all_reduce_cross_replica_sum(out, replication_factor)
```

Comment:
So if I understand this correctly, this collective will only run in the forwards pass (I guess this is an existing-code naming issue that I failed to pick up on before, as I would expect `all_reduce_cross_replica_sum` to be "consistent" and run in both passes by default).

So this & `repcolcol_sharded_matmul` run under the assumption "replicated tensors have replicated gradients", so in this case there's no need to do it in the backwards pass. Is this right? In which case, worth a comment to that effect.

Reply:
> I would expect all_reduce_cross_replica_sum to be "consistent" and run in both passes by default
Why would you expect this? The gradient of `all_reduce_cross_replica_sum` is a copy.

> So this & repcolcol_sharded_matmul run under the assumption "replicated tensors have replicated gradients", so in this case there's no need to do it in the backwards pass. Is this right? In which case, worth a comment to that effect.

Yes, I think this is correct.`repcolcol_sharded_matmul` and `rowcolrep_sharded_matmul` are functions to support the two types of matrix multiplication in the megatron-style tensor parallel case. The left hand block is `repcolcol` and the right hand block is `colrowrep`. The output of the grad op of `rowcolrep_sharded_matmul` should be row-sharded (so no all-reduce needed in the backward pass), and the output of the grad op of `repcolcol_sharded_matmul` should be replicated, (so all-reduce is needed in the backward pass).

<img width="651" alt="Screenshot 2023-03-14 at 12 05 42" src="https://user-images.githubusercontent.com/9449909/224995939-12663f74-f225-4491-b0d0-65f7e2ff889a.png">

Reply:
I believe the way it's implemented makes sense,